### PR TITLE
Encoder: add support for UR types other than 'bytes'

### PR DESCRIPTION
**Pull request by icota**
An attempt to fix https://github.com/dspicher/ur-rs/issues/8. Test is implementing `crypto-request` UR type from [here](https://github.com/BlockchainCommons/crypto-commons/blob/master/Docs/ur-99-request-response.md).

### CODE REVIEW

**Comment by dspicher** [comment]
Hey @icota thanks a lot for the PR, looks interesting! I just fixed the codecov integration, can you rebase please?

**Comment by dspicher** [comment]
The test doesn't check for the same UR as in the [spec](https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2021-001-request.md#example-request-and-response). Also the hexlified transaction id is a different one, maybe that is the reason?

**Comment by icota** [comment]
Looks like I pasted the wrong link, sorry! It's actually [this doc](https://github.com/BlockchainCommons/crypto-commons/blob/master/Docs/ur-99-request-response.md). 
I've amended the test comment and the first comment in the PR.

File: src/ur.rs [hunk 1]
```diff
@@ -74,7 +84,9 @@ impl Decoder {
 #[cfg(test)]
 mod tests {
     use super::*;
+    use hex;
```
Comment (dspicher) [inline src/ur.rs:87]: this can be removed

File: src/ur.rs [hunk 2]
```diff
@@ -84,7 +96,7 @@ mod tests {
     #[test]
     fn test_single_part_ur() {
         let ur = make_message_ur(50, "Wolf");
-        let encoded = Encoder::encode(&ur).unwrap();
+        let encoded = Encoder::encode(&ur, "bytes".to_string()).unwrap();
```
Comment (dspicher) [inline src/ur.rs:99]: `.to_string()` can be removed if you change `encode` as per the other comment

File: src/ur.rs [hunk 3]
```diff
@@ -94,7 +106,7 @@ mod tests {
     #[test]
     fn test_ur_encoder() {
         let ur = make_message_ur(256, "Wolf");
-        let mut encoder = Encoder::new(&ur, 30).unwrap();
+        let mut encoder = Encoder::new(&ur, 30, "bytes".to_string()).unwrap();
```
Comment (dspicher) [inline src/ur.rs:109]: `.to_string()` can be removed if you change `new` as per the other comment

File: src/ur.rs [hunk 4]
```diff
@@ -122,10 +134,44 @@ mod tests {
...
+
+        // 1. Top level is a map
+        let mut top_level_map = BTreeMap::new();
+        top_level_map.insert(Value::Integer(1), uuid);
+        top_level_map.insert(Value::Integer(2), crypto_seed);
+        let top_level = Value::Map(top_level_map);
+
+        let data = serde_cbor::to_vec(&top_level).unwrap();
+
+        let e = Encoder::encode(&data, "crypto-request".to_string()).unwrap();
```
Comment (dspicher) [inline src/ur.rs:166]: `.to_string()` can be removed if you change `encode` as per the other comment

File: src/ur.rs [hunk 5]
```diff
@@ -122,10 +134,44 @@ mod tests {
...
+        let expected = "ur:crypto-request/oeadtpdagdaobncpftlnylfgfgmuztihbawfsgrtflaotaadwkoyadtaaohdhdcxvsdkfgkepezepefrrffmbnnbmdvahnptrdtpbtuyimmemweootjshsmhlunyeslnameyhsdi";
+        assert_eq!(expected, e);
+    }
+
     #[test]
     fn test_multipart_ur() {
         let ur = make_message_ur(32767, "Wolf");
-        let mut encoder = Encoder::new(&ur, 1000).unwrap();
+        let mut encoder = Encoder::new(&ur, 1000, "bytes".to_string()).unwrap();
```
Comment (dspicher) [inline src/ur.rs:174]: `.to_string()` can be removed if you change `new` as per the other comment

**Review by dspicher — CHANGES_REQUESTED** [review]

**Comment by dspicher** [comment]
> Looks like I pasted the wrong link, sorry! It's actually [this doc](https://github.com/BlockchainCommons/crypto-commons/blob/master/Docs/ur-99-request-response.md). I've amended the test comment and the first comment in the PR.

Excellent, thanks. Only some minor stuff left.

File: src/ur.rs [hunk 6]
```diff
@@ -122,10 +134,44 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_ur_encoder_bc_crypto_request() {
+        // https://github.com/BlockchainCommons/crypto-commons/blob/master/Docs/ur-99-request-response.md
```
Comment (dspicher) [inline src/ur.rs:139]: Can we make this more stable by using 
https://github.com/BlockchainCommons/crypto-commons/blob/67ea252f4a7f295bb347cb046796d5b445b3ad3c/Docs/ur-99-request-response.md
?

**Comment by icota** [comment]
Thanks for the review @dspicher! I believe I have implemented all of your suggestions.

**Review by dspicher — APPROVED** [review]

**Comment by dspicher** [comment]
LGTM 👍 thanks again for the PR!

**Pull request merged**